Documentation/contributing/making-changes.rst

```diff
@@ -0,0 +1,256 @@
+   sign up.
+   |br|
+   |br|
+
+
+#. Fork the Project
+
+   Visit both these links and hit the Fork button in the upper right of the page:
+
+   * `NuttX <https://github.com/apache/incubator-nuttx/>`_
+   * `NuttX Apps <https://github.com/apache/incubator-nuttx-apps/>`_
+
+
+#. Change the Git Remotes
+
+   The git repositories in your project are currently connected to the official NuttX repositories, but you don't
+   have permission to push software there. But you can push them to your forks, and from there create Pull Requests
+   if you want to send them to the NuttX project.
+
+   First, remove the current remote, ``origin`` (we'll add it back later):
+
+    .. code-block:: bash
+
+       $ cd nuttx/
+       $ # display the remote
+       $ git remote -v
+
+   You should see something like this:
+
+    .. code-block:: bash
+
+       origin	https://github.com/apache/incubator-nuttx.git
+
+   Now, on the GitHub web page for your forked ``incubator-nuttx`` project, copy the clone url – get it by hitting the
+   green ``Clone or Download`` button in the upper right. Then do this:
+
+    .. code-block:: bash
+
+       $ git remote rm origin
+       $ git remote add origin <your forked incubator-nuttx project clone url>
+       $ git remote add upstream https://github.com/apache/incubator-nuttx.git
+
+   Do the same for your forked ``incubator-nuttx-apps`` project:
+
+    .. code-block:: bash
+
+       $ cd ../apps
+       $ git remote rm origin
+       $ git remote add origin <your forked incubator-nuttx-apps project clone url>
+       $ git remote add upstream https://github.com/apache/incubator-nuttx-apps.git
+
+
+#. Create a Local Git Branch
+
+   Now you can create local git branches and push them to GitHub:
+
+    .. code-block:: bash
+
+       $ git checkout -b test/my-new-branch
+       $ git push
+
+
+Git Workflow With an Upstream Repository
+----------------------------------------
+
+The main NuttX git repository is called an "upstream" repository - this is because it's the main source of truth, and
+its changes flow downstream to people who've forked that repository, like us.
+
+Working with an upstream repo is a bit more complex, but it's worth it since you can submit fixes and features
+to the main NuttX repos. One of the things you need to do regularly is keep your local repo in sync
+with the upstream. I work with a local branch, make changes, pull new software from the upstream and merge it in,
+maybe doing that several times. Then when everything works, I get my branch ready to do a Pull Request. Here's how:
+
+#. Fetch upstream changes and merge into my local master:
+
+    .. code-block:: bash
+
+       $ git checkout master
+       $ git fetch upstream
+       $ git merge upstream/master
+       $ git push
+
+#. Merge my local master with my local branch:
+
+    .. code-block:: bash
+
+       $ git checkout my-local-branch
+       $ git merge master
+       $ git push
```

Comment:
This also is better with a rebase.

Reply:
@Ouss4 Will you post the commands for the rebase flow that you are proposing?

Reply:
They are [here](https://cwiki.apache.org/confluence/display/NUTTX/Code+Contribution+Workflow#CodeContributionWorkflow-BeforeSubmittingYourChanges) with some explanation.
I can provide any other explanation if necessary.

Reply:
Rebase is much better for this. 

I was wondering why we show 
`git rev-list` you can just supply the upstream branch to `git rebase`. We also list the super long form option for force. No one is going to type that.

Reply:
@btashton Will you post the commands for the rebase flow, or point me at a page that specifies them? I'll try the flow and update the instructions.

Reply:
```
git fetch upstream
git rebase upstream/master
git push -u <your remote> -f
```

The `-u` is optional if you already have the tracking branch configured, I just do it to not make the mistake of pushing to upstream where I have permissions.

Reply:
While this is not something for external contributors, but for maintainers (I think having a separate maintainers guide would be good later), I have this set for my nuttx repos:
<pre>
❯ git remote -v
fork	git@github.com:v01d/incubator-nuttx (fetch)
fork	git@github.com:v01d/incubator-nuttx (push)
origin	https://github.com/apache/incubator-nuttx (fetch)
origin	no_push (push)
</pre>
(note origin in my case is upstream)

Reply:
> Rebase is much better for this.
> 
> I was wondering why we show
> `git rev-list` you can just supply the upstream branch to `git rebase`. We also list the super long form option for force. No one is going to type that.

The long version is not the same as the simple force. Supposeddly it is better practice to use the "with-lease" option when force pushing the result of rebase. But I'm not sure if there's a difference when there's only one person accessing a branch.

Reply:
>I was wondering why we show
git rev-list

rev-list or any other count gives the number of commits the branch has.  That number is then used for interactive (-i) rebasing.

> We also list the super long form option for force. No one is going to type that.

Maybe no one is going to type that, but that's not another form of -f.  It's different from -f. --force-with-lease won't let you force push if there are commits upstream that you are going to delete.

Reply:
> >I was wondering why we show
> git rev-list
> 
> rev-list or any other count gives the number of commits the branch has.  That number is then used for interactive (-i) rebasing.
> 

You don't need to do that you just supply the branch as the argument to rebase (or interactive rebase).

Reply:
For just rebase, yes, this is how it's explained. But for interactive rebase, I don't see how that would work.  I'm trying but can't get it to work.  I'm getting "noop" in the rebase "editor".

Reply:
This is what I have done for years so I'm confident that it works. There is no difference in how the ref is resolved interactive or not.

```
git checkout mybranch
git fetch upstream
git rebase upstream/master
```

http://kensheedlo.com/essays/why-you-should-use-a-rebase-workflow/

Reply:
I think we are not talking about the same thing.

For the "just" rebase part, we agree.  You just supply the branch name.  You rebase one branch ontop of the other.
Interactive rebasing is used to change to commits within the same branch.  In the doc's example (here or in Confluence) is used to squash some commits.  It can also be used to delete, change or re-word (change description of) commits.
To do so, we need a range of commits from the HEAD of the branch, thus the HEAD~N.

Reply:
`git rebase -i master` works the same. HEAD~N is just another way to reference a commit (it is not a range)

Reply:
I said a range _from the HEAD_.  I didn't mean that HEAD~N is a range.  HEAD~N is the Nth commit behind HEAD.
The current branch's top commit is HEAD.
When you are "checkedout" to master and you do `git rebase -i master`, that means you are trying to rebase from HEAD to HEAD, thus the "noop".

Reply:
> `git rebase -i master` works the same. HEAD~N is just another way to reference a commit (it is not a range)

Exactly `rev-list` is doing what rebase it already doing when given a branch name.

Reply:
> I said a range _from the HEAD_.  I didn't mean that HEAD ~ N is a range.  HEAD~N is the Nth commit behind HEAD.
> The current branch's top commit is HEAD.
> When you are "checkedout" to master and you do `git rebase -i master`, that means you are trying to rebase from HEAD to HEAD, thus the "noop".

No you are on your branch and interactive rebase against master. This will show all your commits in the interactive flow.

Reply:
> I said a range _from the HEAD_. I didn't mean that HEAD ~ N is a range. HEAD~N is the Nth commit behind HEAD.
> The current branch's top commit is HEAD.
> When you are "checkedout" to master and you do `git rebase -i master`, that means you are trying to rebase from HEAD to HEAD, thus the "noop".

I understand, but the instructions are for a feature branch, not for master. So master is assumed to be the branch point of the feature branch. There shouldn't be any work on master on NuttX (much less any rebasing)

Reply:
> There shouldn't be any work on master on NuttX (much less any rebasing)

Yes, "master" was used just as example.

Ok, now I got you.  You are actually doing both the rebasing on top of master and any other interactive rebase operation you want to do at the same time.
I personally don't use both at the same time.  I squash or drop commits on the branch and not rebase ontop of master until I'm done.
(and no, I don't use rev-list, there are enough bash/zsh plugins that give the number of commits.)

But I'll concide if you guys find it better this way.

Reply:
@Ouss4 @btashton I added the rebase flow listed by @btashton as the first option. I left the other way as a second option. Do you have objections to leaving it in?

Reply:
@btashton I removed the alternative flow. I guess I personally need to switch to this flow too. 🙂